packages/collision/index.ts

```diff
@@ -117,115 +417,256 @@ interface ThrowIfForbiddenOrAmbiguous {
-                (existingMetadata === undefined ||
-                  (existingMetadata !== method &&
-                    (!isArray(existingMetadata) || existingMetadata.indexOf(method) === -1)))
-              );
-            })
-            .forEach(setMethodCallback);
-        });
+function prepareSettings(opts: CollisionSettings | DeprecatedCollisionSettings): CollisionSettings | null {
+  if (!opts) {
+    return null;
+  }
+  const { defer = [] } = (opts || {}) as DeprecatedCollisionSettings;
+  const defaultMethodsSettings = {
+    map: [],
+    reduce: [],
+    reduceThis: [],
+    mapAsync: [],
+    reduceAsync: [],
+    reduceThisAsync: [],
+    allow: [],
+    forbid: [],
+    forbidAll: false,
+  };
+  const defaultInitializersSettings = {};
+  const settings: CollisionSettings = {
+    methods: merge({ ...defaultMethodsSettings }, { map: defer }, opts.methods),
+    initializers: merge({ ...defaultInitializersSettings }, opts.initializers),
+  };
 
-      setMethodsMetadata(opts, methodsMetadata);
-    }
+  return settings;
+}
+
+function validateHasGetSetArguments(domain: string, itemName?: string): void {
+  const dio = domainIsObject(domain);
+  if (dio && !itemName) {
+    throw new Error(`Domain ${domain} requires an item name for aggregates`);
   }
-};
+  if (!dio && itemName) {
+    throw new Error(`Domain ${domain} does not need an item name for aggregates`);
+  }
+}
 
 /**
  * TODO
  */
 const Collision = compose({
-  deepConfiguration: { Collision: { defer: [], forbid: [] } },
+  deepConfiguration: { Collision: {} },
   staticProperties: {
-    collisionSetup(this: Stamp | undefined, opts: CollisionSettings): CollisionStamp {
+    collisionSetup(this: Stamp | undefined, opts: CollisionSettings | DeprecatedCollisionSettings): CollisionStamp {
       return (this?.compose ? this : Collision).compose({
-        deepConfiguration: { Collision: opts },
+        deepConfiguration: { Collision: prepareSettings(opts) },
       }) as CollisionStamp;
     },
     collisionSettingsReset(this: CollisionStamp): CollisionStamp {
+      const settings = getSettings(this.compose, 'initializers') as CollisionSettingsInitializers;
+
+      // Resetting asynchronous initializers will break everything.
+      // Workaround: Use a composer
+      if (settings.async) {
+        throw new Error(
+          'An attempt was made to reset collision settings when the stamp has asynchronous initializers enabled. Use a composer to reset collision settings and properly handle these initializers.'
+        );
+      }
+
       return this.collisionSetup(null);
     },
     collisionProtectAnyMethod(this: CollisionStamp, opts: CollisionSettings): CollisionStamp {
-      return this.collisionSetup(assign({}, opts, { forbidAll: true }) as CollisionSettings);
+      return this.collisionSetup(merge({}, opts, { methods: { forbidAll: true } }) as CollisionSettings);
+    },
+    hasAggregates(this: CollisionStamp, domain: string, itemName?: string) {
```

Comment:
Utility stamps (like the Collision one) tend to give its metadata (like statics) unique names. Typically, the name of the stamp is added to the static method names (not mandatory though). I.e. I'd rather rename these methods:
`hasAggregates` -> `hasCollisionAggregates` or `collisionHasAggregates`.

Reply:
I was able to remove the PropertyMap extension with a few extra functions to retrieve things. You'll see that in the next commit. Will making a new PR for [typescript-PaRt-two](https://github.com/stampit-org/stamp/tree/typescript-PaRt-two) unnecessarily delay access of this stuff to anyone? Since it's essentially ready for master (small details like docs remain), if you decide to merge it into master I'm happy to PR a merge of it into [typescript-PaRt-two](https://github.com/stampit-org/stamp/tree/typescript-PaRt-two). Likewise, if we have to start a new PR for it to go into [typescript-PaRt-two](https://github.com/stampit-org/stamp/tree/typescript-PaRt-two) then can do.

Reply:
Let’s keep this PR as is and mature it.  Don’t worry about the other branch yet

Reply:
@koresar Regarding the has/get/set methods. Naming slipped my mind during implementation. Fixed in next commit.

@PopGoesTheWza Ok